File: server/genesis/runFundingCarryLab.py
#!/usr/bin/env python3
import csv
import io
import json
import math
import os
import statistics
import sys
import urllib.error
import urllib.request
import zipfile
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime, timezone
from pathlib import Path

VERSION = "funding_carry_lab_v2_binance_vision_oos"
BASE = "https://data.binance.vision/data"
SYMBOLS = [x.strip() for x in os.getenv("GENESIS_FUNDING_SYMBOLS", "BTCUSDT,ETHUSDT,SOLUSDT,BNBUSDT,XRPUSDT").split(",") if x.strip()]
MONTHS_BACK = max(4, min(12, int(os.getenv("GENESIS_FUNDING_MONTHS", "6"))))
SIGNAL_LOOKBACK = max(2, int(os.getenv("GENESIS_FUNDING_SIGNAL_LOOKBACK", "3")))
MIN_TRAILING_FUNDING_BPS = float(os.getenv("GENESIS_MIN_TRAILING_FUNDING_BPS", "0.25"))
HORIZONS = [int(x) for x in os.getenv("GENESIS_FUNDING_HORIZONS", "3,6,12").split(",") if x.strip().isdigit() and int(x) > 0]
SPOT_TAKER_BPS_PER_SIDE = float(os.getenv("GENESIS_FUNDING_SPOT_TAKER_BPS", "10"))
PERP_TAKER_BPS_PER_SIDE = float(os.getenv("GENESIS_FUNDING_PERP_TAKER_BPS", "5"))
SLIPPAGE_RESERVE_BPS = float(os.getenv("GENESIS_FUNDING_SLIPPAGE_RESERVE_BPS", "4"))
ROUND_TRIP_COST_BPS = 2 * (SPOT_TAKER_BPS_PER_SIDE + PERP_TAKER_BPS_PER_SIDE) + SLIPPAGE_RESERVE_BPS
OUT = Path(sys.argv[sys.argv.index("--out") + 1]) if "--out" in sys.argv else Path("quant-evidence/funding-carry-lab-latest.json")
USER_AGENT = "GenesisHQ-Research/1.0 (+public-market-data)"


def roundn(value, digits=4):
    try:
        value = float(value)
        return round(value, digits) if math.isfinite(value) else None
    except (TypeError, ValueError):
        return None


def mean(values):
    values = [float(x) for x in values if x is not None and math.isfinite(float(x))]
    return sum(values) / len(values) if values else None


def stdev(values):
    values = [float(x) for x in values if x is not None and math.isfinite(float(x))]
    return statistics.stdev(values) if len(values) >= 2 else None


def normalize_ts(value):
    x = int(float(value))
    while x > 10**14:  # tolerate micro/nanosecond archive timestamps
        x //= 1000
    return x


def completed_months(count):
    now = datetime.now(timezone.utc)
    y, m = now.year, now.month
    out = []
    for _ in range(count):
        m -= 1
        if m == 0:
            y -= 1
            m = 12
        out.append(f"{y:04d}-{m:02d}")
    return list(reversed(out))


def download_csv_from_zip(url):
    req = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
    try:
        with urllib.request.urlopen(req, timeout=30) as response:
            payload = response.read()
    except urllib.error.HTTPError as exc:
        if exc.code == 404:
            return []
        raise RuntimeError(f"http_{exc.code}:{url}") from exc
    with zipfile.ZipFile(io.BytesIO(payload)) as zf:
        names = [name for name in zf.namelist() if name.lower().endswith(".csv")]
        if not names:
            raise RuntimeError(f"zip_without_csv:{url}")
        raw = zf.read(names[0]).decode("utf-8-sig", errors="replace")
    return list(csv.reader(io.StringIO(raw)))


def numeric(value):
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def parse_kline_rows(rows):
    out = []
    for row in rows:
        if len(row) < 5:
            continue
        ts = numeric(row[0])
        close = numeric(row[4])
        if ts is None or close is None or close <= 0:
            continue
        out.append((normalize_ts(ts), float(close)))
    return out


def parse_funding_rows(rows):
    if not rows:
        return []
    header = [x.strip().lower() for x in rows[0]]
    has_header = any(x in {"calc_time", "funding_time", "fundingrate", "funding_rate", "last_funding_rate"} for x in header)
    out = []
    if has_header:
        idx_time = next((i for i, x in enumerate(header) if x in {"calc_time", "funding_time", "fundingtime"}), None)
        idx_rate = next((i for i, x in enumerate(header) if x in {"fundingrate", "funding_rate", "last_funding_rate"}), None)
        if idx_time is None or idx_rate is None:
            return []
        body = rows[1:]
        for row in body:
            if len(row) <= max(idx_time, idx_rate):
                continue
            ts, rate = numeric(row[idx_time]), numeric(row[idx_rate])
            if ts is None or rate is None:
                continue
            out.append((normalize_ts(ts), float(rate)))
        return out

    # Older archive variants are headerless. First field is timestamp; select the
    # small-magnitude numeric field after it (interval hours is usually 8, symbol is text).
    for row in rows:
        if len(row) < 2:
            continue
        ts = numeric(row[0])
        if ts is None:
            continue
        candidates = [numeric(x) for x in row[1:]]
        candidates = [x for x in candidates if x is not None and abs(x) < 0.1]
        if not candidates:
            continue
        out.append((normalize_ts(ts), float(candidates[-1])))
    return out


def urls(symbol, month):
    return {
        "funding": f"{BASE}/futures/um/monthly/fundingRate/{symbol}/{symbol}-fundingRate-{month}.zip",
        "spot": f"{BASE}/spot/monthly/klines/{symbol}/1h/{symbol}-1h-{month}.zip",
        "perp": f"{BASE}/futures/um/monthly/klines/{symbol}/1h/{symbol}-1h-{month}.zip",
    }


def fetch_symbol(symbol, months):
    tasks = []
    for month in months:
        for kind, url in urls(symbol, month).items():
            tasks.append((kind, month, url))
    results = {"funding": [], "spot": [], "perp": []}
    errors = []
    with ThreadPoolExecutor(max_workers=min(8, len(tasks))) as pool:
        futures = {pool.submit(download_csv_from_zip, url): (kind, month, url) for kind, month, url in tasks}
        for future in as_completed(futures):
            kind, month, url = futures[future]
            try:
                rows = future.result()
                parsed = parse_funding_rows(rows) if kind == "funding" else parse_kline_rows(rows)
                results[kind].extend(parsed)
            except Exception as exc:
                errors.append(f"{kind}:{month}:{exc}")
    for kind in results:
        results[kind] = sorted(dict(results[kind]).items())
    return results, errors


def latest_price_at(series, timestamp):
    lo, hi = 0, len(series) - 1
    best = None
    while lo <= hi:
        mid = (lo + hi) // 2
        if series[mid][0] <= timestamp:
            best = series[mid]
            lo = mid + 1
        else:
            hi = mid - 1
    if best is None or timestamp - best[0] > 2 * 60 * 60 * 1000:
        return None
    return best[1]


def align(data):
    spot, perp = data["spot"], data["perp"]
    rows = []
    for ts, rate in data["funding"]:
        spot_px = latest_price_at(spot, ts)
        perp_px = latest_price_at(perp, ts)
        if spot_px and perp_px and spot_px > 0 and perp_px > 0:
            rows.append({"t": ts, "rate": rate, "rateBps": rate * 10000.0, "spot": spot_px, "perp": perp_px})
    return rows


def trade_stats(trades):
    pnl = [float(x["netBps"]) for x in trades if math.isfinite(float(x["netBps"]))]
    wins = [x for x in pnl if x > 0]
    losses = [x for x in pnl if x < 0]
    gp, gl = sum(wins), abs(sum(losses))
    ev, sd = mean(pnl), stdev(pnl)
    curve = peak = max_dd = 0.0
    for value in pnl:
        curve += value
        peak = max(peak, curve)
        max_dd = max(max_dd, peak - curve)
    return {
        "trades": len(pnl),
        "samples": len(pnl),
        "expectancyBps": roundn(ev),
        "profitFactor": roundn(gp / gl) if gl > 0 else (99 if gp > 0 else None),
        "tStat": roundn(ev / (sd / math.sqrt(len(pnl)))) if ev is not None and sd and sd > 0 else None,
        "winRate": roundn(len(wins) / len(pnl)) if pnl else None,
        "maxDrawdownPct": roundn(max_dd / 100.0),
        "totalNetBps": roundn(sum(pnl)),
    }


def simulate(rows, horizon, start=0, end=None):
    end = len(rows) if end is None else min(end, len(rows))
    trades = []
    i = max(start + SIGNAL_LOOKBACK, SIGNAL_LOOKBACK)
    while i + horizon < end:
        signal = mean([row["rateBps"] for row in rows[i - SIGNAL_LOOKBACK:i]])
        current = rows[i]
        if signal is None or signal <= MIN_TRAILING_FUNDING_BPS or current["rateBps"] <= 0:
            i += 1
            continue
        exit_index = i + horizon
        exit_row = rows[exit_index]
        funding_bps = sum(row["rateBps"] for row in rows[i + 1:exit_index + 1])
        spot_return_bps = math.log(exit_row["spot"] / current["spot"]) * 10000.0
        short_perp_return_bps = -math.log(exit_row["perp"] / current["perp"]) * 10000.0
        basis_pnl_bps = spot_return_bps + short_perp_return_bps
        gross_bps = basis_pnl_bps + funding_bps
        net_bps = gross_bps - ROUND_TRIP_COST_BPS
        trades.append({
            "entryTime": current["t"], "exitTime": exit_row["t"], "horizonFundingEvents": horizon,
            "trailingFundingSignalBps": roundn(signal), "fundingReceivedBps": roundn(funding_bps),
            "spotReturnBps": roundn(spot_return_bps), "shortPerpReturnBps": roundn(short_perp_return_bps),
            "basisPnlBps": roundn(basis_pnl_bps), "grossBps": roundn(gross_bps),
            "roundTripCostBps": roundn(ROUND_TRIP_COST_BPS), "netBps": roundn(net_bps),
        })
        i = exit_index + 1
    return trades


def choose_horizon(rows, train_end):
    ranked = []
    for horizon in HORIZONS:
        trades = simulate(rows, horizon, 0, train_end)
        stats = trade_stats(trades)
        score = stats["expectancyBps"] * math.sqrt(stats["trades"]) if stats["trades"] >= 8 and stats["expectancyBps"] is not None else -math.inf
        ranked.append((score, horizon, stats))
    ranked.sort(reverse=True, key=lambda x: x[0])
    return ranked[0] if ranked else None


def passes_oos(validation, holdout):
    return (
        validation["trades"] >= 8 and holdout["trades"] >= 8
        and (validation["expectancyBps"] or -math.inf) > 0 and (holdout["expectancyBps"] or -math.inf) > 0
        and (validation["profitFactor"] or 0) >= 1.05 and (holdout["profitFactor"] or 0) >= 1.10
        and (holdout["tStat"] or -math.inf) >= 0.75
        and (holdout["maxDrawdownPct"] if holdout["maxDrawdownPct"] is not None else math.inf) <= 12
    )


def evaluate_symbol(symbol, months):
    data, errors = fetch_symbol(symbol, months)
    rows = align(data)
    if len(rows) < 120:
        return None, errors + [f"insufficient_aligned:{len(rows)}"]
    train_end, valid_end = int(len(rows) * 0.60), int(len(rows) * 0.80)
    selected = choose_horizon(rows, train_end)
    if selected is None or not math.isfinite(selected[0]):
        return {
            "symbol": symbol, "alignedFundingEvents": len(rows), "selectedHorizon": selected[1] if selected else None,
            "train": selected[2] if selected else trade_stats([]), "validation": trade_stats([]), "holdout": trade_stats([]),
            "oosTrades": 0, "passedOos": False, "evidenceQuality": 0.35, "evidenceStatus": "TRAIN_SAMPLE_INSUFFICIENT",
        }, errors
    _, horizon, train_stats = selected
    validation_trades = simulate(rows, horizon, train_end, valid_end)
    holdout_trades = simulate(rows, horizon, valid_end, len(rows))
    validation, holdout = trade_stats(validation_trades), trade_stats(holdout_trades)
    passed = passes_oos(validation, holdout)
    oos_trades = validation["trades"] + holdout["trades"]
    evidence_quality = min(0.95, 0.72 + min(oos_trades, 60) / 260) if passed else min(0.59, 0.40 + min(oos_trades, 40) / 220)
    latest = rows[-1]
    recent_mean = mean([x["rateBps"] for x in rows[-SIGNAL_LOOKBACK:]])
    return {
        "symbol": symbol,
        "archiveMonths": months,
        "alignedFundingEvents": len(rows),
        "selectedHorizon": horizon,
        "latestFundingBps": roundn(latest["rateBps"]),
        "recentMeanFundingBps": roundn(recent_mean),
        "latestBasisBps": roundn(math.log(latest["perp"] / latest["spot"]) * 10000.0),
        "train": train_stats,
        "validation": validation,
        "holdout": holdout,
        "oosTrades": oos_trades,
        "passedOos": passed,
        "evidenceQuality": roundn(evidence_quality),
        "evidenceStatus": "DELTA_NEUTRAL_CARRY_OOS_PASS" if passed else "DELTA_NEUTRAL_CARRY_OOS_FAIL",
        "recentHoldoutTrades": holdout_trades[-5:],
    }, errors


def main():
    months = completed_months(MONTHS_BACK)
    candidates, errors = [], []
    for symbol in SYMBOLS:
        try:
            candidate, symbol_errors = evaluate_symbol(symbol, months)
            errors.extend([f"{symbol}:{x}" for x in symbol_errors])
            if candidate:
                candidates.append(candidate)
        except Exception as exc:
            errors.append(f"{symbol}:{exc}")
    candidates.sort(key=lambda x: (bool(x.get("passedOos")), x.get("holdout", {}).get("expectancyBps") if x.get("holdout", {}).get("expectancyBps") is not None else -math.inf), reverse=True)
    best = candidates[0] if candidates else None
    output = {
        "ok": True,
        "version": VERSION,
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "mode": "RESEARCH_ONLY",
        "paperOnly": True,
        "executionAuthority": False,
        "liveLocked": True,
        "liveOrders": False,
        "methodology": {
            "source": "Official Binance Vision monthly archives: USD-M funding + spot 1h + perpetual 1h",
            "archiveMonths": months,
            "symbols": SYMBOLS,
            "split": "60% train / 20% validation / 20% holdout",
            "signal": f"prior {SIGNAL_LOOKBACK} funding events mean > {MIN_TRAILING_FUNDING_BPS} bps and current funding > 0",
            "horizonsFundingEvents": HORIZONS,
            "horizonSelection": "train only; frozen for validation and holdout",
            "trade": "long spot + short USD-M perpetual, equal-notional log-return approximation",
            "pnl": "spot return - perp return + funding received - explicit entry/exit fees and slippage reserve",
            "spotTakerBpsPerSide": SPOT_TAKER_BPS_PER_SIDE,
            "perpTakerBpsPerSide": PERP_TAKER_BPS_PER_SIDE,
            "slippageReserveBps": SLIPPAGE_RESERVE_BPS,
            "roundTripCostBps": ROUND_TRIP_COST_BPS,
            "negativeFundingPolicy": "FAIL_CLOSED: no short-spot/long-perp because borrow cost/availability is not modeled",
            "limitation": "Historical delta-neutral approximation; no margin/liquidation-path model or fee-tier discounts; no real positions are opened.",
        },
        "testedSymbols": len(candidates),
        "oosPassCount": sum(1 for x in candidates if x.get("passedOos")),
        "candidates": candidates,
        "best": best,
        "sleeve": {
            "sleeveKey": "FUNDING_CARRY",
            "engineVersion": VERSION,
            "samples": best.get("oosTrades", 0),
            "expectancyBps": best.get("holdout", {}).get("expectancyBps"),
            "profitFactor": best.get("holdout", {}).get("profitFactor"),
            "tStat": best.get("holdout", {}).get("tStat"),
            "maxDrawdownPct": best.get("holdout", {}).get("maxDrawdownPct"),
            "evidenceQuality": best.get("evidenceQuality", 0),
            "paperCapitalEligible": bool(best.get("passedOos")),
        } if best else None,
        "errors": errors[:50],
        "invariants": {"signsTransactions": False, "broadcastsTransactions": False, "unlocksLive": False},
    }
    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(json.dumps(output, indent=2) + "\n", encoding="utf-8")
    print(json.dumps({
        "version": VERSION,
        "testedSymbols": output["testedSymbols"],
        "oosPassCount": output["oosPassCount"],
        "best": {"symbol": best["symbol"], "selectedHorizon": best.get("selectedHorizon"), "validation": best.get("validation"), "holdout": best.get("holdout"), "passedOos": best.get("passedOos")} if best else None,
        "errors": len(output["errors"]),
    }))


if __name__ == "__main__":
    main()

File: server/genesis/runFundingCarryLab.mjs
import { mkdir, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';

const VERSION = 'funding_carry_lab_v2_causal_spot_close_oos';
const FUTURES_BASE = process.env.BINANCE_FUTURES_BASE || 'https://fapi.binance.com';
const SPOT_BASE = process.env.BINANCE_SPOT_BASE || 'https://data-api.binance.vision/api/v3';
const SYMBOLS = (process.env.GENESIS_FUNDING_SYMBOLS || 'BTCUSDT,ETHUSDT,SOLUSDT,BNBUSDT,XRPUSDT,DOGEUSDT,LINKUSDT').split(',').map((x) => x.trim()).filter(Boolean);
const FUNDING_LIMIT = Math.max(120, Math.min(1000, Number(process.env.GENESIS_FUNDING_HISTORY_LIMIT || 400)));
const SPOT_BAR_LIMIT = Math.max(1200, Math.min(5000, Number(process.env.GENESIS_FUNDING_SPOT_BARS || 3500)));
const SIGNAL_LOOKBACK = Math.max(2, Number(process.env.GENESIS_FUNDING_SIGNAL_LOOKBACK || 3));
const MIN_TRAILING_FUNDING_BPS = Number(process.env.GENESIS_MIN_TRAILING_FUNDING_BPS || 0.25);
const HOLD_HORIZONS = (process.env.GENESIS_FUNDING_HORIZONS || '3,6,12').split(',').map(Number).filter((x) => Number.isInteger(x) && x > 0);
const SPOT_TAKER_BPS_PER_SIDE = Number(process.env.GENESIS_FUNDING_SPOT_TAKER_BPS || 10);
const PERP_TAKER_BPS_PER_SIDE = Number(process.env.GENESIS_FUNDING_PERP_TAKER_BPS || 5);
const SLIPPAGE_RESERVE_BPS = Number(process.env.GENESIS_FUNDING_SLIPPAGE_RESERVE_BPS || 4);
const ROUND_TRIP_COST_BPS = 2 * (SPOT_TAKER_BPS_PER_SIDE + PERP_TAKER_BPS_PER_SIDE) + SLIPPAGE_RESERVE_BPS;
const OUT = process.argv.includes('--out') ? process.argv[process.argv.indexOf('--out') + 1] : 'quant-evidence/funding-carry-lab-latest.json';

const HOUR_MS = 60 * 60 * 1000;
const round = (x, d = 4) => Number.isFinite(Number(x)) ? Number(Number(x).toFixed(d)) : null;
const mean = (a) => a.length ? a.reduce((s, x) => s + x, 0) / a.length : null;
const std = (a) => { if (a.length < 2) return null; const m = mean(a); return Math.sqrt(a.reduce((s, x) => s + (x - m) ** 2, 0) / (a.length - 1)); };
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

async function fetchJson(url) {
  const response = await fetch(url, { signal: AbortSignal.timeout(12_000) });
  if (!response.ok) {
    const body = await response.text().catch(() => '');
    throw new Error(`http_${response.status}:${url}:${body.slice(0, 100)}`);
  }
  return response.json();
}

async function fundingHistory(symbol) {
  const q = new URLSearchParams({ symbol, limit: String(FUNDING_LIMIT) });
  const rows = await fetchJson(`${FUTURES_BASE}/fapi/v1/fundingRate?${q}`);
  if (!Array.isArray(rows)) throw new Error(`invalid_funding:${symbol}`);
  return rows
    .map((row) => ({
      t: Number(row.fundingTime),
      rate: Number(row.fundingRate),
      rateBps: Number(row.fundingRate) * 10_000,
      perp: Number(row.markPrice),
    }))
    .filter((row) => Number.isFinite(row.t) && row.t > 0 && Number.isFinite(row.rate) && row.perp > 0)
    .sort((a, b) => a.t - b.t);
}

export function spotRowsFromKlines(rows = []) {
  return rows.map((row) => {
    const openTime = Number(row?.[0]);
    const close = Number(row?.[4]);
    const closeTime = Number(row?.[6]);
    if (!Number.isFinite(openTime) || !Number.isFinite(closeTime) || closeTime <= openTime || !(close > 0)) return null;
    return { t: closeTime, close, openTime };
  }).filter(Boolean).sort((a, b) => a.t - b.t);
}

async function spotHistory(symbol) {
  const all = [];
  let endTime = Date.now();
  while (all.length < SPOT_BAR_LIMIT) {
    const limit = Math.min(1000, SPOT_BAR_LIMIT - all.length);
    const q = new URLSearchParams({ symbol, interval: '1h', limit: String(limit), endTime: String(endTime) });
    const page = await fetchJson(`${SPOT_BASE}/klines?${q}`);
    if (!Array.isArray(page) || !page.length) break;
    all.unshift(...page);
    if (page.length < limit) break;
    endTime = Number(page[0][0]) - 1;
    await sleep(25);
  }
  const unique = [...new Map(all.map((row) => [Number(row[0]), row])).values()]
    .sort((a, b) => Number(a[0]) - Number(b[0]))
    .slice(-SPOT_BAR_LIMIT);
  return spotRowsFromKlines(unique);
}

export function nearestSpot(spots, t) {
  if (!spots.length) return null;
  let lo = 0, hi = spots.length - 1, best = null;
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    if (spots[mid].t <= t) { best = spots[mid]; lo = mid + 1; } else hi = mid - 1;
  }
  if (!best || t - best.t > 2 * HOUR_MS) return null;
  return best.close;
}

export function align(funding, spots) {
  return funding.map((row) => ({ ...row, spot: nearestSpot(spots, row.t) })).filter((row) => row.spot > 0);
}

function tradeStats(trades) {
  const pnl = trades.map((x) => x.netBps).filter(Number.isFinite);
  const wins = pnl.filter((x) => x > 0), losses = pnl.filter((x) => x < 0);
  const gp = wins.reduce((s, x) => s + x, 0), gl = Math.abs(losses.reduce((s, x) => s + x, 0));
  const ev = mean(pnl), sd = std(pnl);
  let curve = 0, peak = 0, maxDdBps = 0;
  for (const x of pnl) { curve += x; peak = Math.max(peak, curve); maxDdBps = Math.max(maxDdBps, peak - curve); }
  return {
    trades: pnl.length,
    samples: pnl.length,
    expectancyBps: round(ev, 4),
    profitFactor: gl > 0 ? round(gp / gl, 4) : (gp > 0 ? 99 : null),
    tStat: ev != null && sd > 0 ? round(ev / (sd / Math.sqrt(pnl.length)), 4) : null,
    winRate: pnl.length ? round(wins.length / pnl.length, 4) : null,
    maxDrawdownPct: round(maxDdBps / 100, 4),
    totalNetBps: round(pnl.reduce((s, x) => s + x, 0), 4),
  };
}

function simulate(rows, horizon, start = 0, end = rows.length) {
  const trades = [];
  let i = Math.max(start + SIGNAL_LOOKBACK, SIGNAL_LOOKBACK);
  while (i + horizon < end) {
    const signalWindow = rows.slice(i - SIGNAL_LOOKBACK, i);
    const trailingFundingBps = mean(signalWindow.map((x) => x.rateBps));
    const current = rows[i];
    // Direction is deliberately one-sided: positive/persistent funding only.
    // Negative funding would require spot borrowing; borrow cost is not assumed free.
    if (!(trailingFundingBps > MIN_TRAILING_FUNDING_BPS && current.rateBps > 0)) { i++; continue; }
    const exitIndex = i + horizon;
    const exit = rows[exitIndex];
    const fundingReceivedBps = rows.slice(i + 1, exitIndex + 1).reduce((sum, x) => sum + x.rateBps, 0);
    const spotReturnBps = Math.log(exit.spot / current.spot) * 10_000;
    const shortPerpReturnBps = -Math.log(exit.perp / current.perp) * 10_000;
    const basisPnlBps = spotReturnBps + shortPerpReturnBps;
    const grossBps = basisPnlBps + fundingReceivedBps;
    const netBps = grossBps - ROUND_TRIP_COST_BPS;
    trades.push({
      entryTime: current.t,
      exitTime: exit.t,
      horizonFundingEvents: horizon,
      trailingFundingSignalBps: round(trailingFundingBps, 4),
      fundingReceivedBps: round(fundingReceivedBps, 4),
      spotReturnBps: round(spotReturnBps, 4),
      shortPerpReturnBps: round(shortPerpReturnBps, 4),
      basisPnlBps: round(basisPnlBps, 4),
      grossBps: round(grossBps, 4),
      roundTripCostBps: round(ROUND_TRIP_COST_BPS, 4),
      netBps: round(netBps, 4),
    });
    i = exitIndex + 1;
  }
  return trades;
}

function chooseTrainHorizon(rows, trainEnd) {
  const ranked = HOLD_HORIZONS.map((horizon) => {
    const trades = simulate(rows, horizon, 0, trainEnd);
    const stats = tradeStats(trades);
    const score = stats.trades >= 8 && stats.expectancyBps != null ? stats.expectancyBps * Math.sqrt(stats.trades) : -Infinity;
    return { horizon, stats, score };
  }).sort((a, b) => b.score - a.score);
  return ranked[0] ?? null;
}

function passesOos(validation, holdout) {
  return validation.trades >= 8 && holdout.trades >= 8
    && (validation.expectancyBps ?? -Infinity) > 0 && (holdout.expectancyBps ?? -Infinity) > 0
    && (validation.profitFactor ?? 0) >= 1.05 && (holdout.profitFactor ?? 0) >= 1.10
    && (holdout.tStat ?? -Infinity) >= 0.75
    && (holdout.maxDrawdownPct ?? Infinity) <= 12;
}

async function evaluateSymbol(symbol) {
  const [funding, spots] = await Promise.all([fundingHistory(symbol), spotHistory(symbol)]);
  const rows = align(funding, spots);
  if (rows.length < 100) throw new Error(`insufficient_aligned_history:${symbol}:${rows.length}`);
  const trainEnd = Math.floor(rows.length * 0.60);
  const validEnd = Math.floor(rows.length * 0.80);
  const selected = chooseTrainHorizon(rows, trainEnd);
  if (!selected || !Number.isFinite(selected.score)) {
    return { symbol, alignedFundingEvents: rows.length, selectedHorizon: selected?.horizon ?? null, train: selected?.stats ?? tradeStats([]), validation: tradeStats([]), holdout: tradeStats([]), passedOos: false, evidenceQuality: 0.35, evidenceStatus: 'TRAIN_SAMPLE_INSUFFICIENT' };
  }
  const validationTrades = simulate(rows, selected.horizon, trainEnd, validEnd);
  const holdoutTrades = simulate(rows, selected.horizon, validEnd, rows.length);
  const validation = tradeStats(validationTrades), holdout = tradeStats(holdoutTrades);
  const passedOos = passesOos(validation, holdout);
  const oosTrades = validation.trades + holdout.trades;
  const evidenceQuality = passedOos ? Math.min(0.95, 0.72 + Math.min(oosTrades, 60) / 260) : Math.min(0.59, 0.40 + Math.min(oosTrades, 40) / 220);
  const latest = rows.at(-1);
  const recentMeanFundingBps = round(mean(rows.slice(-SIGNAL_LOOKBACK).map((x) => x.rateBps)), 4);
  return {
    symbol,
    alignedFundingEvents: rows.length,
    selectedHorizon: selected.horizon,
    latestFundingBps: round(latest?.rateBps, 4),
    recentMeanFundingBps,
    latestBasisBps: latest ? round(Math.log(latest.perp / latest.spot) * 10_000, 4) : null,
    train: selected.stats,
    validation,
    holdout,
    oosTrades,
    passedOos,
    evidenceQuality: round(evidenceQuality, 4),
    evidenceStatus: passedOos ? 'DELTA_NEUTRAL_CARRY_OOS_PASS' : 'DELTA_NEUTRAL_CARRY_OOS_FAIL',
    recentHoldoutTrades: holdoutTrades.slice(-5),
  };
}

async function main() {
  const candidates = [];
  const errors = [];
  for (const symbol of SYMBOLS) {
    try { candidates.push(await evaluateSymbol(symbol)); }
    catch (error) { errors.push(`${symbol}:${String(error?.message || error)}`); }
  }
  candidates.sort((a, b) => Number(b.passedOos) - Number(a.passedOos) || (b.holdout?.expectancyBps ?? -Infinity) - (a.holdout?.expectancyBps ?? -Infinity));
  const best = candidates[0] ?? null;
  const output = {
    ok: true,
    version: VERSION,
    generatedAt: new Date().toISOString(),
    mode: 'RESEARCH_ONLY',
    paperOnly: true,
    executionAuthority: false,
    liveLocked: true,
    liveOrders: false,
    methodology: {
      source: 'Binance public USD-M funding history + public spot 1h klines',
      causalPriceAlignment: 'Spot close is timestamped by Binance kline closeTime (column 6) and is eligible only when closeTime <= fundingTime.',
      priorEvidenceCompatibility: 'funding_carry_lab_v1_delta_neutral_oos used spot close values timestamped at kline open and must be treated as causally invalid.',
      symbols: SYMBOLS,
      fundingHistoryLimit: FUNDING_LIMIT,
      split: '60% train / 20% validation / 20% holdout',
      signal: `prior ${SIGNAL_LOOKBACK} funding events mean > ${MIN_TRAILING_FUNDING_BPS} bps and current funding > 0`,
      horizonsFundingEvents: HOLD_HORIZONS,
      horizonSelection: 'train only; frozen for validation and holdout',
      trade: 'long spot + short perpetual, equal notional approximation',
      pnl: 'spot log return - perp log return + received funding - explicit round-trip fees/slippage',
      spotTakerBpsPerSide: SPOT_TAKER_BPS_PER_SIDE,
      perpTakerBpsPerSide: PERP_TAKER_BPS_PER_SIDE,
      slippageReserveBps: SLIPPAGE_RESERVE_BPS,
      roundTripCostBps: ROUND_TRIP_COST_BPS,
      negativeFundingPolicy: 'FAIL_CLOSED: no short-spot/long-perp carry because borrow availability/cost is not modeled',
      limitation: 'Historical delta-neutral approximation; ignores margin liquidation path and venue/account-specific fee discounts. No real positions are opened.',
    },
    testedSymbols: candidates.length,
    oosPassCount: candidates.filter((x) => x.passedOos).length,
    candidates,
    best,
    sleeve: best ? {
      sleeveKey: 'FUNDING_CARRY',
      engineVersion: VERSION,
      samples: best.oosTrades ?? 0,
      expectancyBps: best.holdout?.expectancyBps ?? null,
      profitFactor: best.holdout?.profitFactor ?? null,
      tStat: best.holdout?.tStat ?? null,
      maxDrawdownPct: best.holdout?.maxDrawdownPct ?? null,
      evidenceQuality: best.evidenceQuality ?? 0,
      paperCapitalEligible: best.passedOos === true,
    } : null,
    errors: errors.slice(0, 30),
    invariants: { signsTransactions: false, broadcastsTransactions: false, unlocksLive: false },
  };
  await mkdir(dirname(OUT), { recursive: true });
  await writeFile(OUT, `${JSON.stringify(output, null, 2)}\n`);
  console.log(JSON.stringify({ version: VERSION, testedSymbols: output.testedSymbols, oosPassCount: output.oosPassCount, best: best ? { symbol: best.symbol, selectedHorizon: best.selectedHorizon, validation: best.validation, holdout: best.holdout, passedOos: best.passedOos } : null, errors: output.errors.length }));
}

if (process.argv[1]?.endsWith('runFundingCarryLab.mjs')) {
  main().catch((error) => { console.error(error); process.exitCode = 1; });
}
